### PR TITLE
Make ZStream.unfold safe to pull again

### PR DESCRIPTION
**Pull request by vasilmkd**
@iravid

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -3066,19 +3066,25 @@ object ZStream extends Serializable {
...
-        .flatMap(f0)
-        .foldM(
-          e => Pull.fail(e),
-          opt =>
-            opt match {
-              case Some((a, s)) => ref.set(s) *> Pull.emit(a)
-              case None         => Pull.end
-            }
-        )
+        done <- Ref.make(false).toManaged_
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: So we can use `Pull.memoizeEnd` here right?

```
Pull.memoizeEnd {
  Ref.make(s).toManaged_.map {
    // previous code here
  }
}
```

**Comment by vasilmkd** [comment]
I started disliking `memoizeEnd` because it also captures the error, so it's not completely correct in this case. We should resume normally after a potential error in `f0`, not end.

**Comment by iravid** [comment]
So let's fix it, by not capturing the error, or delete it?

**Comment by vasilmkd** [comment]
I vote for deleting

**Pull request merged**